src/core/announcements/factories.ts

```diff
@@ -22,19 +22,53 @@ export type Announcement = TypedAnnouncement<AnnouncementType>;
 /**
  * TypedAnnouncement: an Announcement with a particular AnnouncementType
  */
-export interface TypedAnnouncement<T extends AnnouncementType> {
+export type TypedAnnouncement<T extends AnnouncementType> = {
   announcementType: T;
   fromId: DSNPUserId;
-  createdAt: BigInt;
-}
+  createdAt: bigint;
+} & (TombstoneFields | BroadcastFields | ReplyFields | ReactionFields | GraphChangeFields | ProfileFields);
+
+type TombstoneFields = {
+  announcementType: AnnouncementType.Tombstone;
+  targetAnnouncementType: AnnouncementType;
+  targetSignature: HexString;
+};
+
+type BroadcastFields = {
+  announcementType: AnnouncementType.Broadcast;
+  contentHash: HexString;
+  url: string;
+};
+
+type ReplyFields = {
+  announcementType: AnnouncementType.Reply;
+  contentHash: HexString;
+  inReplyTo: DSNPAnnouncementURI;
+  url: string;
+};
+
+type ReactionFields = {
+  announcementType: AnnouncementType.Reaction;
+  emoji: string;
+  inReplyTo: DSNPAnnouncementURI;
+};
+
+type GraphChangeFields = {
+  announcementType: AnnouncementType.GraphChange;
+  changeType: DSNPGraphChangeType;
+  objectId: DSNPUserId;
+};
+
+type ProfileFields = {
+  announcementType: AnnouncementType.Profile;
+  contentHash: HexString;
+  url: string;
+};
```

Comment:
why is this preferable over the extended interfaces?

Reply:
It let's the type resolve from just the AnnouncementType generic correctly. The interface extensions are weaker and wouldn't be able to resolve the fields properly.